core/src/repair_service.rs

```diff
@@ -345,6 +359,14 @@ impl RepairService {
         }
     }
 
+    fn insert_new_and_filter_old_nonces(new_nonces: Vec<Nonce>, nonces: &mut HashMap<Nonce, u64>) {
+        let now = timestamp();
+        for nonce in new_nonces {
+            nonces.insert(nonce, now);
+        }
+        nonces.retain(|_nonce, timestamp| now.saturating_sub(*timestamp) < MAX_NONCE_AGE_MS);
```

Comment:
The `retain` thing is traversing the entire hashmap for pruning, which might be slow.
Another idea is to use `LruCache<Nonce, u64>` which prunes by itself.
`verify_repair` can also check that if the nonce has expired, and if so remove it from the cache as well.

Reply:
True, good idea.